src/ethereum/prague/vm/precompiled_contracts/bls12_381/__init__.py

```diff
@@ -0,0 +1,441 @@
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+Precompile for BLS12-381 curve operations.
+"""
+from typing import Tuple, Union
+
+from py_ecc.bls12_381.bls12_381_curve import (
+    FQ,
+    FQ2,
+    b,
+    b2,
+    curve_order,
+    is_on_curve,
+    multiply,
+)
+from py_ecc.optimized_bls12_381.optimized_curve import FQ as OPTIMIZED_FQ
+from py_ecc.optimized_bls12_381.optimized_curve import FQ2 as OPTIMIZED_FQ2
+from py_ecc.typing import Point2D
+
+from ethereum.base_types import U256, Bytes
+
+from ....vm.memory import buffer_read
+from ...exceptions import InvalidParameter
+
+P = FQ.field_modulus
+
+K_DISCOUNT = {
+    1: 1200,
+    2: 888,
+    3: 764,
+    4: 641,
+    5: 594,
+    6: 547,
+    7: 500,
+    8: 453,
+    9: 438,
+    10: 423,
+    11: 408,
+    12: 394,
+    13: 379,
+    14: 364,
+    15: 349,
+    16: 334,
+    17: 330,
+    18: 326,
+    19: 322,
+    20: 318,
+    21: 314,
+    22: 310,
+    23: 306,
+    24: 302,
+    25: 298,
+    26: 294,
+    27: 289,
+    28: 285,
+    29: 281,
+    30: 277,
+    31: 273,
+    32: 269,
+    33: 268,
+    34: 266,
+    35: 265,
+    36: 263,
+    37: 262,
+    38: 260,
+    39: 259,
+    40: 257,
+    41: 256,
+    42: 254,
+    43: 253,
+    44: 251,
+    45: 250,
+    46: 248,
+    47: 247,
+    48: 245,
+    49: 244,
+    50: 242,
+    51: 241,
+    52: 239,
+    53: 238,
+    54: 236,
+    55: 235,
+    56: 233,
+    57: 232,
+    58: 231,
+    59: 229,
+    60: 228,
+    61: 226,
+    62: 225,
+    63: 223,
+    64: 222,
+    65: 221,
+    66: 220,
+    67: 219,
+    68: 219,
+    69: 218,
+    70: 217,
+    71: 216,
+    72: 216,
+    73: 215,
+    74: 214,
+    75: 213,
+    76: 213,
+    77: 212,
+    78: 211,
+    79: 211,
+    80: 210,
+    81: 209,
+    82: 208,
+    83: 208,
+    84: 207,
+    85: 206,
+    86: 205,
+    87: 205,
+    88: 204,
+    89: 203,
+    90: 202,
+    91: 202,
+    92: 201,
+    93: 200,
+    94: 199,
+    95: 199,
+    96: 198,
+    97: 197,
+    98: 196,
+    99: 196,
+    100: 195,
+    101: 194,
+    102: 193,
+    103: 193,
+    104: 192,
+    105: 191,
+    106: 191,
+    107: 190,
+    108: 189,
+    109: 188,
+    110: 188,
+    111: 187,
+    112: 186,
+    113: 185,
+    114: 185,
+    115: 184,
+    116: 183,
+    117: 182,
+    118: 182,
+    119: 181,
+    120: 180,
+    121: 179,
+    122: 179,
+    123: 178,
+    124: 177,
+    125: 176,
+    126: 176,
+    127: 175,
+    128: 174,
+}
+
+MAX_DISCOUNT = 174
+MULTIPLIER = 1000
+
+
+def bytes_to_G1(data: Bytes) -> Point2D:
+    """
+    Decode 128 bytes to a G1 point. Does not perform sub-group check.
+
+    Parameters
+    ----------
+    data :
+        The bytes data to decode.
+
+    Returns
+    -------
+    point : Point2D
+        The G1 point.
+
+    Raises
+    ------
+    InvalidParameter
+        Either a field element is invalid or the point is not on the curve.
+    """
+    assert len(data) == 128
```

Comment:
These are utility functions. Calling them without the appropriate length of bytes is invalid behaviour. The assertion is a reminder for any function calling this utility in the future to pass the right size. As far as the current implmentation goes, the size checks are already performed in the calling function with the appropriate error raise. See `bls12_381_g1.bls12_g1_add` for example.

However, I would be ok with changing this to a `if...raise`

Reply:
Yeah, I can accept that argument. I'll leave this up to your judgment.

My reasoning was that since `bytes_to_G1` accepts a `Bytes` as a parameter, it should handle every instance of that type. If this took, for example, a `Bytes128` it would make sense to use an assertion.

Reply:
Sounds good. Updated